src/connection.rs

```diff
@@ -0,0 +1,41 @@
+use std::ffi::c_void;
+
+use libc::{size_t, EIO};
+use rustls::Session;
+
+use crate::error::rustls_io_error;
+use crate::io::{CallbackReader, CallbackWriter, ReadCallback, WriteCallback};
+
+// Call Session::read_tls, providing an &mut dyn Write implemented with a C callback.
+pub(crate) fn read_tls(
+    session: &mut dyn Session,
```

Comment:
Could this be `read_tls<S: Session>(session: &mut S, ...)` to avoid dynamic dispatch? It seems like this and `write_tls` will be on the hot path so maybe it's worth specializing for `ClientSession` and `ServerSession`? Perhaps the time spent on dynamic dispatch is irrelevant since you're waiting on IO all the time anyway.

Reply:
yeah, I think dynamic dispatch will wind up not being a big deal. And in #99 dynamic dispatch becomes even more a part of the library in a way that I think would be somewhat painful to remove.